### ISSUE TITLE
file_packager: Add suport for glob or loading file list from file

### ISSUE THREAD

**Issue by KivApple** [emscripten-core/emscripten#18366]
Sometimes you might not want to pack all files from a specific folder, but only some. In this case glob pattern (there is [a function](https://docs.python.org/3/library/glob.html) in the Python standard library for doing this) would be useful like `assets/**/*.png`, but it is not supported by `file_packager`. There is a workaround - you can specify all files explicitly by hand or generate a command line using another script. But in this case you may hit command line length limit on Windows if you have too many files. In such case it would be useful to be able to read file list from text file (e.g. compilers allow you to pass options using text files). But such feature is also missing from `file_packager`.

**Comment by sbc100** [comment]
Both those features sound reasonable to me.

**Comment by debadree25** [comment]
If i understand correctly as per this issue we want to be able to do something like `emcc file.cpp -o file.html --preload-file asset_dir/**/*.png`?

**Comment by debadree25** [comment]
Another question I had at https://github.com/emscripten-core/emscripten/blob/03c61d544cc679d8555f3d1c6887d357bb5ea59d/tools/file_packager.py#L431 
we are setting source and destination path from the arg but when we use globs what should be the dest and source path? like in case of `asset_dir/**/*.png` should asset_dir maybe be the dest path? @sbc100

**Comment by sbc100** [comment]
I think if we you want a different source and dest path then you use @ symbol in the argument so `uses_at_notation` would be set in that case.

I agree we would have to come up with a mapping from source the dest that might be a little more complex.  Its not clear how it would work, but I think the @ sign should continue to be the signal that you want a different source and dest.

**Comment by debadree25** [comment]
Yeah, there's no issue when the @ character is there then no issue its when we don't have it, any ideas or inspirations you can think of in this case? I have no experience here so asking 😅

**Comment by sbc100** [comment]
When the @ char is not there, the user is saying that source and dest pathnames are the same.  So shouldn't we just map every file in the glob that create corresponding file with the same name in the dest root.

**Comment by debadree25** [comment]
Ahh I think then if src is glob like we can recursively find all the files and append them as `DataFile` that can work i think

**Comment by sbc100** [comment]
How else would the glob work, if not by recursively finding all the file?    You can use python `glob` or `fnmatch` modules to do the finding, right?

**Comment by debadree25** [comment]
yes yes was a little confused at first 😅😅, will attempt this

**Comment by KivApple** [comment]
I think @ can replace file path part until first glob symbol *.

E. g. `asset_dir/**/*.png@dir` would put everything under `asset_dir` into dir preserving nested directory structure.

**Comment by ChinYikMing** [comment]
Preload or embed necessary files inside a directory can reduce the size of WASM module substantially if there are multiple unnecessary files inside the same directory. In my [case](https://github.com/sysprog21/rv32emu/issues/75), I can reduce a from size 62M to 23M.

I think I can help to solve this issue. May I be the assignee?

**Comment by sbc100** [comment]
By all means, feel free to submit PR for this.

**Comment by sbc100** [comment]
Having said that, cant these use cases also be solves using a 2-phase deployment setup:

1. Copy of the files you want to deploy into a "dist" directory.
2. Pass this "dist" directory to emscripten/file_packager.

**Comment by ChinYikMing** [comment]
> Having said that, cant these use cases also be solves using a 2-phase deployment setup:
> 
> 1. Copy of the files you want to deploy into a "dist" directory.
> 2. Pass this "dist" directory to emscripten/file_packager.

Seems workable! However, I think providing a general mapping option might easier for user (at least for me).

**Comment by sbc100** [comment]
I guess it depends on how much complexity this would add to file packager, vs how it is for users to do this copying/preparation step.   

In the world of open source software its very common to have a `make install` type step that stages all the files necessary for a program.  My current inclination is that this is reasonable thing for toolchainers to do themselves and that we adding more complexity to file packager should be avoided if we can.
